sweep/aggregator.go

```diff
@@ -86,7 +224,19 @@ func (s *SimpleAggregator) ClusterInputs(inputs pendingInputs) []inputCluster {
 	// Since the inputs that we clustered by fee rate don't commit to a
 	// specific locktime, we can try to merge a locktime cluster with a fee
 	// cluster.
-	return zipClusters(lockTimeClusters, feeClusters)
+	clusters := zipClusters(lockTimeClusters, feeClusters)
```

Comment:
The function could be renamed to make it a bit more clear, but what it does today is merged the two lists, preserving the fee rate ordering from the first. In other words, `lockTimeClusters` will have merged into the items from `feeClusters` that are as high as the element in the lock time cluster.

Reply:
I think we need to be careful here as a zip operation (pairwise combination) needs to make sure there is proper index alignment.

Diving into the implementation details it seems that this zipClusters operation isn't actually a zip (at least with the terminology I'm familiar with) because it does two independent sorts by feerate on each vector of inputs. What concerns me further here is that since feerates are not enumerable it means that we can't really reason about the relative densities between the two slices. If we don't have some sort of guarantee on how dense each slice is wrt feerate gain, then a single advancement of the `j` cursor per `i` cursor advancement may not make sense. I'm still trying to wrap my head around what the cluster operations are. Maybe a high level pseudocode description without loops and cursors is needed.

My understanding here is that `clusterByLocktime` allocates a bunch of locktime buckets, into which we are distributing the non-locktime inputs provided they meet some sort of fee distance criteria (still trying to figure out exactly what criteria that is).

> that are as high as the element in the lock time cluster

I don't see any guarantee that there is a single element in the locktime cluster, so maybe it is if the input is at least as high as the cluster average feerate? That would make sense to me because we wouldn't want to drag down the feerate of a cluster that had time sensitivity.

@yyforyongyu can you confirm this interpretation for me and then maybe we can work on making it more obvious in code?

Reply:
IMO the current clustering logic is unnecessarily complicated, I'm in favor of removing `SimpleAggregator` once the `BudgetAggregator` is in place.

Reply:
Right but can we confirm or correct my above interpretation?

Reply:
From what I've learnt from `TestZipClusters`, think once talked about this in a study group - my understanding is that, once clustered by locktimes and fee rate buckets, these clusters are merged if the timelock clusters have a smaller fee rate than the fee rate clusters.

The effect is, for locktime clusters, their fee rates will be increased if the non-locktime clusters have higher ones.